### ISSUE TITLE
Inconsistent reasoner behaviour on reflexive relations

### ISSUE THREAD

**Issue by redstonedude0** [graknlabs/grakn#5649]
## Description

Data inserted by rules does not adhere to same restrictions as data inserted directly (specifically the same entity is allowed to play the same role twice in a relation). Data 'inserted' by rules seems to not adhere to some restrictions placed on regular inserts, specifically the same entity can play the same role twice in any given relation.

## Environment

1. OS (where Grakn server runs): Ubuntu 18.04.4 LTS
2. Grakn version (and platform): Grakn (core?) 1.6.2 ["grakn server version" just says '1.6.2']
3. Grakn client: console
4. Other environment details: fresh/clean keyspace

## Reproducible Steps

Steps to create the smallest reproducible scenario:
1. Start grakn console

2. Paste the following into console

````
define name sub attribute, datatype string;
define friendship sub relation, relates friend;
define person sub entity, has name, plays friend;
define friendship-transitive sub rule, when { (friend: $f1, friend: $f2) isa friendship;(friend: $f2, friend: $f3) isa friendship;}, then {  (friend: $f1, friend: $f3) isa friendship;};
insert $x isa person, has name "Alice";
insert $x isa person, has name "Bob";
insert $x isa person, has name "Charlie";
match $alice isa person, has name "Alice"; $bob isa person, has name "Bob"; insert (friend:$alice,friend:$bob) isa friendship;
match $alice isa person, has name "Alice"; $charlie isa person, has name "Charlie"; insert (friend:$alice,friend:$charlie) isa friendship;
commit
match (friend: $f1, friend: $f2) isa friendship; $f1 has name $f1N; $f2 has name $f2N; get $f1N, $f2N;
````

## Expected Output

I would expect to see
A-B,
A-C,
B-A,
B-C,
C-A,
C-B
as the only 6 results

## Actual Output

I see the above 6, but additionally I see
A-A,
B-B,
C-C
as another 3 results, despite that being one entity playing the "friend" role twice in "friendship", which can't be achieved by directly inserting

**Comment by flyingsilverfin** [comment]
related to #5600

**Comment by flyingsilverfin** [comment]
The targeted solution to this, in conjunction with language behaviors in #5658 which is the result of a discussion with Haikal and James, is for now to fail rule valdiation if the rule head inferring a relation with identical role players does not include a `!=` in the `when`:

```
define friendship-transitive sub rule, 
when {
 (friend: $f1, friend: $f2) isa friendship;
 (friend: $f2, friend: $f3) isa friendship;
}, then {  
 (friend: $f1, friend: $f3) isa friendship;
};
```
would be disallowed as `$f1` and `$f3` play the same role in a relation, and are not explicitly defined to be unequal

We would have to modify to read:

```
define friendship-transitive sub rule, 
when {
 (friend: $f1, friend: $f2) isa friendship;
 (friend: $f2, friend: $f3) isa friendship;
 $f1 != f3;
}, then {  
 (friend: $f1, friend: $f3) isa friendship;
};
```

This is overly verbose for many cases, such as when the `when` declares the variables to have disjoint types. However, this can be implemented in the future with more advanced validation.

**Comment by flyingsilverfin** [comment]
The Above comment has become obsoleted by a furtherd discussion, which has its decision outlined in #5669

**Comment by flyingsilverfin** [comment]
Closed with completion of steps in #5669 . Thanks for reporting!
